kad-dht/README.md

```diff
@@ -0,0 +1,427 @@
+areas:
+
+1. Specialisations and peculiarities of the libp2p implementation.
+2. Actual wire messages.
+3. Other algorithmic or non-standard behaviours worth pointing out.
+
+For everything else that isn't explicitly stated herein, it is safe to assume
+behaviour similar to Kademlia-based libraries.
+
+Code snippets use a Go-like syntax.
+
+## Definitions
+
+### Replication parameter (`k`)
+
+The amount of replication is governed by the replication parameter `k`. The
+default value for `k` is 20.
+
+### Kademlia routing table
+
+The data structure backing this system is a k-bucket routing table, closely
+following the design outlined in the Kademlia paper [0]. The bucket size is
+equal to the replication paramter `k`, and the maximum bucket count matches the
+size of the SHA256 function, i.e. 256 buckets.
+
+### Alpha concurrency parameter (`α`)
+
+The concurrency of node and value lookups are limited by parameter `α`, with a
+default value of 3. This implies that each lookup process can perform no more
+than 3 inflight requests, at any given time.
+
+### Distance
+
+In all cases, the distance between two keys is `XOR(sha256(key1),
+sha256(key2))`.
+
+## DHT operations
+
+The libp2p Kademlia DHT offers the following types of routing operations:
+
+- **Peer routing** - _Finding_ the closest nodes to a given key (`FIND_NODE`).
+- **Value routing** - _Putting_ a value to the nodes closest to the value's key
+  (`PUT_VALUE`) and _getting_ a value by its key from the nodes closest to that
+  key (`GET_VALUE`).
+- **Content routing** - _Adding_ oneself to the list of providers for a given
+  key at the nodes closest to that key (`ADD_PROVIDER`) and _getting_ providers
+  for a given key from the nodes closest to that key (`GET_PROVIDERS`).
+
+In addition the libp2p Kademlia DHT offers the auxiliary _bootstrap_ operation.
+
+### Peer routing
+
+The below is one possible algorithm to find nodes closest to a given key on
+the DHT. Implementations may diverge from this base algorithm as long as they
+continue to adhere to the wire format.
+
+Let's assume we’re looking for nodes closest to key `Key`. We then enter an
+iterative network search.
+
+We keep track of the set of peers we've already queried (`Pq`) and the set of
+next query candidates sorted by distance from `Key` in ascending order (`Pn`).
+At initialization `Pn` is seeded with the `α` peers from our routing table we
+know are closest to `Key`, based on the XOR distance function (see [distance
+definition](#distance)).
+
+Then we loop:
+
+1. > The lookup terminates when the initiator has queried and gotten responses
+   from the k (see [#replication-parameter-k]) closest nodes it has seen.
+
+   (See Kademlia paper [0].)
+
+   The lookup might terminate early in case the local node queried all known
+   nodes, with the number of nodes being smaller than `k`.
+2. Pick as many peers from the candidate peers (`Pn`) as the `α` concurrency
```

Comment:
When do you do the picking? Our implementation does this after a prior request completes and the state is updated.

Reply:
I am not sure I follow. Steps 1 - 4 are executed in a loop, thus picking happens at the very start as well as each time a response is received (3) and the termination criterion (1) is still false. As far as I can tell, this is inline with what you describe above. Does that make sense @petar?

Reply:
You are right... I overlooked. This looks fine.